File: JAVA/Sum of nodes on the longest path from root to leaf node.java
class Solution {
    int maxH, res;
    void solve(Node root, int ht, int sum) {
        if(root == null)
            return;
        
        if(root.left == null && root.right == null) {
            sum += root.data;
            
            if(ht > maxH) {
                maxH = ht;
                res = sum;
            } else if(ht == maxH) 
                res = Math.max(res, sum);
                
            return;
        }
        
        sum += root.data;
        solve(root.left, ht + 1, sum);
        solve(root.right, ht + 1, sum);
    }
    
    int sumOfLongRootToLeafPath(Node root) {
        maxH = res = 0;
        solve(root, 0, 0);
        return res;
    }
}

File: C++/Sum of nodes on the longest path from root to leaf node.cpp
class Solution
{    
public:
    int ans = 0, maxPath = 0;
    
    void solve(Node* root, vector<int>& temp, int sum) {
        if(!root) {
            int n = (int)temp.size();
            
            if(maxPath < n) {
                maxPath = n;
                ans = sum;
            }
            if(maxPath == n) {
               if(sum>ans) ans = sum;
            }
            return;
        }
        temp.push_back(root->data);
        sum += root->data;
        solve(root->left, temp, sum);
        solve(root->right, temp, sum);
        sum -= root->data;
        temp.pop_back();
        return;
    }
    int sumOfLongRootToLeafPath(Node *root)
    {
        vector<int> temp;
        solve(root, temp, 0);
        return ans;
    }
};
